### PR TITLE
Using new Prometheus native histograms for Cilium metrics

### PR DESCRIPTION
**Pull request by pedro-stanaka**
## Summary of changes

I have observed that Cilium is still using traditional native histograms to instrument most of its metrics, in this PR I am enabling native histogram for Cilium instrumentation. More info on native histograms in the [official Prometheus talk](https://promcon.io/2022-munich/talks/native-histograms-in-prometheus/). 

For ops teams, this change should bring no difference since native histograms are only scraped if a feature flag is enabled on Prometheus side, that is, if they are not using the feature their metrics will still be exporter with the same name and using the same bucket configuration of the existing metrics.
If the feature is enabled in the Prometheus scraping Cilium, the metric names will change, users won't see the metrics with suffixes like `_count`, `_sum`, `_bucket` anymore.

The main benefits of using native histogram are that the metrics become "lighter" in cardinality (since it removes the "fake label" `le`) and also becomes much more precise.

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [ ] Thanks for contributing!

Fixes: #issue-number

```release-note
Enabling native histograms for Cilium related metrics to reduce cardinality and improve precision
```

### CODE REVIEW

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -31,6 +31,11 @@ const (
 	// ErrorProxy is the value used to notify errors on Proxy.
 	ErrorProxy = "proxy"
 
+	// HistogramFactor configures the factor used for exponential/native histogram in Prometheus client
+	HistogramFactor = 1.1
```
Comment (tommyp1ckles) [inline pkg/metrics/metrics.go:35]: Where did this factor value come from, how do we know it works for all metrics?

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -31,6 +31,11 @@ const (
...
+	// HistogramMaxBuckets configures the maximum number of buckets that will be exported in Prometheus client
+	HistogramMaxBuckets = 128
```
Comment (tommyp1ckles) [inline pkg/metrics/metrics.go:37]: Same question as above.

**Comment by tommyp1ckles** [comment]
Thanks for the contribution 🙏 😄 

Just trying to understand any possible impact of this better.
I am a bit concerned about the stability of this feature, the [documentation ](https://prometheus.io/docs/prometheus/latest/feature_flags/#native-histograms)mentions that this feature is still quite experimental. Notably, it mentions possibly with including breaking changes for the TSDM format.

I think this would be worth discussing in one of our [community meetings](https://docs.cilium.io/en/latest/community/community/)

**Comment by tommyp1ckles** [comment]
Also, FYI, there's some more Cilium Operator specific metrics living in `operator/metrics/` and `pkg/ipam` 😄

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (pedro-stanaka) [reply pkg/metrics/metrics.go:35]: This controls how many buckets will be created for the exponential/native histogram. The bucket layout configured on the metric, will also be respected.

This is better documented here: 

https://github.com/prometheus/client_golang/blob/main/prometheus/histogram.go#L398-L429

> The smaller
> // NativeHistogramBucketFactor, the more buckets will be used and thus
> 	// the more costly the histogram will become. A generally good trade-off
> 	// between cost and accuracy is a value of 1.1 (each bucket is at most
> 	// 10% wider than the previous one), which will result in each power of
> 	// two divided into 8 buckets (e.g. there will be 8 buckets between 1
> 	// and 2, same as between 2 and 4, and 4 and 8, etc.).
> 	//
> 	// Details about the actually used factor: The factor is calculated as
> 	// 2^(2^n), where n is an integer number between (and including) -8 and
> 	// 4. n is chosen so that the resulting factor is the largest that is
> 	// still smaller or equal to NativeHistogramBucketFactor. Note that the
> 	// smallest possible factor is therefore approx. 1.00271 (i.e. 2^(2^-8)
> 	// ). If NativeHistogramBucketFactor is greater than 1 but smaller than
> 	// 2^(2^-8), then the actually used factor is still 2^(2^-8) even though
> 	// it is larger than the provided NativeHistogramBucketFactor.

File: pkg/metrics/metrics.go (hunk 2 above)
Comment (pedro-stanaka) [reply pkg/metrics/metrics.go:37]: With this new feature, you could endup with unbounded dynamic buckets, since each observation may create a new bucket in the histogram. To avoid problems with memory, we introduce this parameter to limit the amount of buckets exported by the client golang.

Comment (pedro-stanaka) [reply pkg/metrics/metrics.go:37]: Also documented here: https://github.com/prometheus/client_golang/blob/main/prometheus/histogram.go#L442-L466

**Comment by pedro-stanaka** [comment]
> Thanks for the contribution 🙏 😄
> 
> Just trying to understand any possible impact of this better. I am a bit concerned about the stability of this feature, the [documentation ](https://prometheus.io/docs/prometheus/latest/feature_flags/#native-histograms) mentions that this feature is still quite experimental. Notably, it mentions possibly with including breaking changes for the TSDM format.
> 
> Might it be premature to be adding this?
> 
> I think this would be worth discussing in one of our [community meetings](https://docs.cilium.io/en/latest/community/community/)

This feature is currently hidden behind a feature flag in Prometheus, so users of Cilium that are still not using the native histogram feature, are still going to see the normal histograms as they were seeing before.

If they opt-in though, to native histograms they will see the native histogram version of the metric. 

The only problem here is that native histogram metrics change their names, for example, where you had `agent_boot_time_seconds` and metrics like: `agent_boot_time_seconds_bucket`, `agent_boot_time_seconds_sum`, `agent_boot_time_seconds_count`. Now you only have the `agent_boot_time_seconds` and you will need to apply new functions prefixed with histogram, to find out the sum/count of the histogram. Example: `histogram_count(rate(agent_boot_time_seconds[2m]))`.

**Comment by pedro-stanaka** [comment]
I see that you put this PR up for discussion on the next call, so I will try to drop by the meeting and see what folks have to say.

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -1224,22 +1251,26 @@ func CreateConfiguration(metricsEnabled []string) (Configuration, []prometheus.C
...
-				Namespace: Namespace,
-				Subsystem: SubsystemKVStore,
-				Name:      "events_queue_seconds",
-				Help:      "Duration in seconds of time received event was blocked before it could be queued",
-				Buckets:   []float64{.002, .005, .01, .015, .025, .05, .1, .25, .5, .75, 1},
+				Namespace:                      Namespace,
+				Subsystem:                      SubsystemKVStore,
+				Name:                           "events_queue_seconds",
+				Help:                           "Duration in seconds of time received event was blocked before it could be queued",
+				Buckets:                        []float64{.002, .005, .01, .015, .025, .05, .1, .25, .5, .75, 1},
```
Comment (ghost) [inline pkg/metrics/metrics.go:1283]: At the moment some of the histograms specify the buckets explicitly, and others do not. This will make the native histograms inconsistent I think. From the [client_golang docs](https://github.com/prometheus/client_golang/blob/main/prometheus/histogram.go#L390-L395):
```
If Buckets is left as nil or set to a slice of length
zero, it is replaced by default buckets. The default buckets are
DefBuckets if no buckets for a native histogram (see below) are used,
otherwise the default is no buckets. (In other words, if you want to
use both reguler buckets and buckets for a native histogram, you have
to define the regular buckets here explicitly.)
```

So if Prometheus is configured to scrape native histograms, then for metrics with explicit buckets it will scrape both native and traditional histograms, but for others only the native histogram? Am I reading this right? In such case it would be nice to have it consistent IMO, i.e. define the buckets for all metrics explicitly.

Comment (pedro-stanaka) [reply pkg/metrics/metrics.go:1283]: Oh, true. Very good catch. I will set `Buckets` as `DefBuckets` to the histograms that do not have bucket layout configured.

**Comment by lambdanis** [comment]
Thanks @pedro-stanaka for the update.

I know it will be discussed in more details in the community meeting, but if it's a go then the feature should be documented [somewhere here](https://github.com/cilium/cilium/blob/main/Documentation/observability/metrics.rst) I think. And as @tommyp1ckles mentioned - there are Cilium Operator specific histograms, it would be good to have them consistent with Cilium Agent metrics. For Hubble metrics I don't think it's as necessary, they're used differently (although maybe something to consider too).

File: pkg/metrics/metrics.go (hunk 3 above)
Comment (pedro-stanaka) [reply pkg/metrics/metrics.go:1283]: I resolved that now.

File: pkg/metrics/metrics.go (hunk 1 above)
Comment (asauber) [reply pkg/metrics/metrics.go:35]: The choice of these two parameters should depend on the range of expected values for each metric, otherwise you end up with a histogram that is "tail-heavy". That is, the final bucket contains everything, and the histogram is effectively useless.

I think at the least, each histogram should have this configurable, using these values as defaults.

The choice of these parameters is fairly significant, because once we start reporting these metrics with certain bucket sizes we can't change them without affecting the meaning of the metric for all users. We should document typical ranges for each of these metrics, and how these choices have influenced these bucket size parameters.

The final bucket range with these values appears to be 1.1^128, or ~200,000x the minimum value, based on my cursory understanding of the implementation (having implemented histograms in other contexts the past).

**Comment by lambdanis** [comment]
As discussed in the community meeting, it would be good to have a flag and a Helm value to explicitly enable native histograms.

This means more configuration for users who want to use them: users will need to enable native histograms in Prometheus and additionally in Cilium (and in any other metrics exporter that will follow such approach). However, native histograms are highly experimental and bring non-trivial underlying changes: switch to gRPC for scraping, non-stable TSDB format, resources usage in particular in the case when you need to scrape both old and new. Because of that, I imagine users might want to enable native histograms in Prometheus, but only for one particular service, not necessarily Cilium.

File: pkg/ipam/metrics/metrics.go [hunk 4]
```diff
@@ -9,6 +9,7 @@ import (
 	"github.com/prometheus/client_golang/prometheus"
 
 	"github.com/cilium/cilium/operator/metrics"
+	metrics2 "github.com/cilium/cilium/pkg/metrics"
```
Comment (squeed) [inline pkg/ipam/metrics/metrics.go:12]: extremely minor nit: call this pkgMetrics or something similar.

**Comment by squeed** [comment]
This looks really useful, and thanks for enabling this. Of course, native histograms haven't settled yet, but they certainly are very compelling from a server operations perspective.

One thought about interacting with multiple versions of Prometheus: should we add a rule in our ServiceMonitor to drop native histograms, and a corresponding Helm option? (I'm not sure if that's possible, but maybe?) What happens if we expose native histograms to a version of Prometheus that doesn't support it? Will it cause problems and/or metric spam?

In other words, does this need to be behind a feature flag? If so, maybe we can avoid plumbing this all through the codebase and just drop at the ServiceMonitor instead.

Reading the docs, there's nothing explicitly needed for newer versions of Prometheus. In other words, it seems like client-side buckets coexist peacefully with native histograms. If this understanding is incorrect, do let me know.

**Comment by pedro-stanaka** [comment]
> This looks really useful, and thanks for enabling this. Of course, native histograms haven't settled yet, but they certainly are very compelling from a server operations perspective.
> 
> One thought about interacting with multiple versions of Prometheus: should we add a rule in our ServiceMonitor to drop native histograms, and a corresponding Helm option? (I'm not sure if that's possible, but maybe?) What happens if we expose native histograms to a version of Prometheus that doesn't support it? Will it cause problems and/or metric spam?

This would be cool, but it is not possible. If the Prometheus administrator enabled the feature flag "enable-native-histograms" the new metric version will be scraped directly from the target.

> In other words, does this need to be behind a feature flag? If so, maybe we can avoid plumbing this all through the codebase and just drop at the ServiceMonitor instead.

As @lambdanis said, the agreement reached in the community call was that we needed a feature flag from Cillium side.

Something I wonder here is how a Cilium user that can't control the initialization options from Cilium can enable this feature flag? For example, if I want to use Control Plane V2 from Google's GKE. I would have to wait anyway until we decide to remove this from a feature flag.

> Reading the docs, there's nothing explicitly needed for newer versions of Prometheus. In other words, it seems like client-side buckets coexist peacefully with native histograms. If this understanding is incorrect, do let me know.

You are correct.
